File: OneToOneCallSample/app/src/main/java/com/stringee/apptoappcallsample/stringee/common/AudioDeviceResolver.java
package com.stringee.apptoappcallsample.stringee.common;

import com.stringee.common.StringeeAudioManager.AudioDevice;

import java.util.Set;

/** Pure routing policy used to select peripheral and built-in audio devices. */
final class AudioDeviceResolver {
    enum RouteAction {
        OBSERVE_ONLY,
        APPLY_BUILT_IN_ONLY,
        SELECT_DEVICE
    }

    private AudioDeviceResolver() {
    }

    static AudioDevice resolveInitialDevice(boolean videoCall, AudioDevice selectedDevice,
                                            Set<AudioDevice> availableDevices) {
        if (isPeripheral(selectedDevice) && contains(availableDevices, selectedDevice)) {
            return selectedDevice;
        }
        return resolveBuiltInDevice(videoCall, availableDevices);
    }

    static AudioDevice resolveBuiltInDevice(boolean videoCall,
                                            Set<AudioDevice> availableDevices) {
        AudioDevice preferred = videoCall ? AudioDevice.SPEAKER_PHONE : AudioDevice.EARPIECE;
        if (contains(availableDevices, preferred)) {
            return preferred;
        }
        AudioDevice fallback = videoCall ? AudioDevice.EARPIECE : AudioDevice.SPEAKER_PHONE;
        return contains(availableDevices, fallback) ? fallback : AudioDevice.NONE;
    }

    static AudioDevice resolvePreferredBuiltIn(AudioDevice preferred,
                                               Set<AudioDevice> availableDevices) {
        if (isBuiltIn(preferred) && contains(availableDevices, preferred)) {
            return preferred;
        }
        return resolveBuiltInDevice(preferred == AudioDevice.SPEAKER_PHONE, availableDevices);
    }

    static boolean isBuiltIn(AudioDevice device) {
        return device == AudioDevice.SPEAKER_PHONE || device == AudioDevice.EARPIECE;
    }

    static boolean isPeripheral(AudioDevice device) {
        return device == AudioDevice.BLUETOOTH || device == AudioDevice.WIRED_HEADSET;
    }

    static RouteAction resolveRouteAction(boolean userInitiated, AudioDevice device) {
        if (userInitiated) {
            return RouteAction.SELECT_DEVICE;
        }
        return isBuiltIn(device) ? RouteAction.APPLY_BUILT_IN_ONLY
                : RouteAction.OBSERVE_ONLY;
    }

    private static boolean contains(Set<AudioDevice> availableDevices, AudioDevice device) {
        return availableDevices != null && availableDevices.contains(device);
    }
}

File: Kotlin-OneToOneCallSample/app/src/main/java/com/stringee/kotlin_onetoonecallsample/stringee/common/AudioDeviceResolver.kt
package com.stringee.kotlin_onetoonecallsample.stringee.common

import com.stringee.common.StringeeAudioManager.AudioDevice

/** Pure routing policy used to select peripheral and built-in audio devices. */
internal object AudioDeviceResolver {
    @JvmStatic fun resolveInitialDevice(
        videoCall: Boolean, selectedDevice: AudioDevice?,
        availableDevices: MutableSet<AudioDevice?>?
    ): AudioDevice? {
        if (isPeripheral(selectedDevice) && contains(availableDevices, selectedDevice)) {
            return selectedDevice
        }
        return resolveBuiltInDevice(videoCall, availableDevices)
    }

    @JvmStatic fun resolveBuiltInDevice(
        videoCall: Boolean,
        availableDevices: MutableSet<AudioDevice?>?
    ): AudioDevice {
        val preferred = if (videoCall) AudioDevice.SPEAKER_PHONE else AudioDevice.EARPIECE
        if (contains(availableDevices, preferred)) {
            return preferred
        }
        val fallback = if (videoCall) AudioDevice.EARPIECE else AudioDevice.SPEAKER_PHONE
        return if (contains(availableDevices, fallback)) fallback else AudioDevice.NONE
    }

    @JvmStatic fun resolvePreferredBuiltIn(
        preferred: AudioDevice?,
        availableDevices: MutableSet<AudioDevice?>?
    ): AudioDevice? {
        if (isBuiltIn(preferred) && contains(availableDevices, preferred)) {
            return preferred
        }
        return resolveBuiltInDevice(preferred == AudioDevice.SPEAKER_PHONE, availableDevices)
    }

    @JvmStatic fun isBuiltIn(device: AudioDevice?): Boolean {
        return device == AudioDevice.SPEAKER_PHONE || device == AudioDevice.EARPIECE
    }

    @JvmStatic fun isPeripheral(device: AudioDevice?): Boolean {
        return device == AudioDevice.BLUETOOTH || device == AudioDevice.WIRED_HEADSET
    }

    @JvmStatic fun resolveRouteAction(userInitiated: Boolean, device: AudioDevice?): RouteAction {
        if (userInitiated) {
            return RouteAction.SELECT_DEVICE
        }
        return if (isBuiltIn(device))
            RouteAction.APPLY_BUILT_IN_ONLY
        else
            RouteAction.OBSERVE_ONLY
    }

    private fun contains(
        availableDevices: MutableSet<AudioDevice?>?,
        device: AudioDevice?
    ): Boolean {
        return availableDevices != null && availableDevices.contains(device)
    }

    internal enum class RouteAction {
        OBSERVE_ONLY,
        APPLY_BUILT_IN_ONLY,
        SELECT_DEVICE
    }
}
